### ISSUE TITLE
Publishing data on internet is slow

### ISSUE THREAD

**Issue by vikaschamp** [alanxz/SimpleAmqpClient#175]
Hello Alanxz ,
I am using SimpleAmqpClient lib in vs2017 .  Which i installed from nuget package in vs2017.
It's publishing data very slow on internet .
It's working fine on localhost (local machine).

Below is my code 

```cpp
AmqpClient::Channel::ptr_t connection = AmqpClient::Channel::Create("192.36.56.89",9009, "finsoltech", "pass123!", "/");
const std::string body = "Testing msg";
AmqpClient::BasicMessage b(body);
const boost::shared_ptr<AmqpClient::BasicMessage> msg{ new AmqpClient::BasicMessage(body) };
std::string queName = connection->DeclareQueue("myqueue",false, false , true , true);
std::cout << "queue has been declare named as" << queName << std::endl;
connection->DeclareExchange("myexchange", "direct", false, false, false);
std::cout << "Exchange has been declare"<< std::endl;
connection->BindQueue("myqueue", "myexchange", "");
std::cout << "Exchange has been bind" << std::endl;
while (true)
	{
	connection->BasicPublish("myexchange", "", msg, false, false);
	std::cout << "Publishing data" << i++ << std::endl;
	}
```

Please suggest me , what should i do.

Thanks in advance.

**Comment by alanxz** [comment]
What do you mean by slow? 

The AMQP protocol was designed to run over high-bandwidth LANs, you won't get the best performance over links with limited bandwidth or those that have a high latency.

**Comment by vikaschamp** [comment]
Thanks for reply Sir.

Sir my bandwidth is good . Because i am using rabbit mq  client in c# it's giving good performance .
When i am executing above code on localhost machine it's publishing 1000 msg/second.
But when i send to remote machine it's publishing aprrox 10 msg per second .
I recheck my bandwidth also , it's good .

May you tell me if there any change in above code .

Thanks

**Comment by alanxz** [comment]
SimpleAmqpClient publishes messages synchronously that is: when Channel::BasicPublish is called it sends the message, then waits for an acknowledgement from the broker before returning from the function (unlikely the C# client which does not wait for acknowledgement, acknowledgements are processed synchronously).

This is a design decision on how SimpleAmqpClient is implemented in that it trades simplicity when it comes to the using the API vs performance.

**Comment by vikaschamp** [comment]
Yes Sir, you said right . I was thinking this only.
Sir , is there any asynchronous method for publishing method in SimpleAmqpClient library .

**Comment by alanxz** [comment]
> is there any asynchronous method for publishing method in SimpleAmqpClient library .

Currently no.

**Comment by vikaschamp** [comment]
So what should i do sir , May you suggest me any other library. I tried for AMQP-CPP library also but i am getting runtime excxeption there.

**Comment by alanxz** [comment]
Use https://github.com/alanxz/rabbitmq-c, it does not force this synchronous behavior.

**Comment by vikaschamp** [comment]
I have intalled this also from nuget package in vs2017.
But how to use , i mean there is no example for implementation .
How to creat channel , declare queue , and publish messege.
Is it same as SimpleAmqpClient implementation.

**Comment by vikaschamp** [comment]
https://github.com/alanxz/rabbitmq-c,  , this library has asynchronous publishing method ?

**Comment by vikaschamp** [comment]
Thank you Sir , I am using https://github.com/alanxz/rabbitmq-c this library till now it's working fine.

But some time it's choking .

**Comment by vikaschamp** [comment]
Closes #175
